extension/k8sleaderelector/extension.go

```diff
@@ -42,20 +49,43 @@ type leaderElectionExtension struct {
 
 	onStartedLeading []StartCallback
 	onStoppedLeading []StopCallback
+
+	onStartLeadingChan chan StartCallback
+
+	mu sync.Mutex
 }
 
 // If the receiver sets a callback function then it would be invoked when the leader wins the election
 func (lee *leaderElectionExtension) startedLeading(ctx context.Context) {
+	lee.onStartLeadingChan = make(chan StartCallback, 1)
+
 	for _, callback := range lee.onStartedLeading {
 		callback(ctx)
 	}
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				callback, ok := <-lee.onStartLeadingChan
+				if !ok {
+					return
+				}
+				if callback != nil {
+					callback(ctx)
+				}
+			}
+		}
+	}()
 }
 
 // If the receiver sets a callback function then it would be invoked when the leader loss the election
 func (lee *leaderElectionExtension) stoppedLeading() {
 	for _, callback := range lee.onStoppedLeading {
 		callback()
 	}
+	close(lee.onStartLeadingChan)
```

Comment:
Why do you close the channel here?

Reply:
The channel is closed so that no new callbacks are pushed, for a scenario when the election is lost, then new callbacks are not pushed. These will be appended into the list and executed when its elected leader again.